### PR TITLE
3.11

### PR DESCRIPTION
**Pull request by ghickman**
Fixes: #2310

### CODE REVIEW

**Comment by ghickman** [comment]
@bloodearnest – I've copied reports' method for getting python from deadsnakes.  I don't understand why I had to take the python dependencies out of `docker/dependencies.txt` and while it currently works I'm keen to get this right before anything else copies it!  Could you take a look and see if this is the right way to do it?

File: docker/Dockerfile [hunk 1]
```diff
@@ -13,6 +13,14 @@ FROM ghcr.io/opensafely-core/base-docker:22.04 as base-python
 # docker clean up that deletes that cache on every apt install
 RUN rm -f /etc/apt/apt.conf.d/docker-clean
 
+# ensure fully working base python3.11 installation using deadsnakes ppa
+# see: https://gist.github.com/tiran/2dec9e03c6f901814f6d1e8dad09528e
+# use space efficient utility from base image
+RUN --mount=type=cache,target=/var/cache/apt \
+    echo "deb http://ppa.launchpad.net/deadsnakes/ppa/ubuntu jammy main" > /etc/apt/sources.list.d/deadsnakes-ppa.list &&\
+    /usr/lib/apt/apt-helper download-file 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf23c5a6cf475977595c89f51ba6932366a755776' /etc/apt/trusted.gpg.d/deadsnakes.asc &&\
```
Comment (bloodearnest) [inline docker/Dockerfile:21]:
```suggestion
    /usr/lib/apt/apt-helper download-file 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf23c5a6cf475977595c89f51ba6932366a755776' /etc/apt/trusted.gpg.d/deadsnakes.asc
```

File: docker/Dockerfile [hunk 2]
```diff
@@ -13,6 +13,14 @@ FROM ghcr.io/opensafely-core/base-docker:22.04 as base-python
...
+    root/docker-apt-install.sh python3.11 python3.11-venv python3.11-distutils
```
Comment (bloodearnest) [inline docker/Dockerfile:22]:
```suggestion
```

**Review by bloodearnest — COMMENTED** [review]
I've add some suggested changes, but I can't do it all, as I can't suggest changes to deleted lines.

Basically, add the ppa as one RUN command, don't install python there.

Then, put those 3.11 packages into dependencies.txt.  This puts all the actual installs into 1 RUN command and thus one layer, rather than spreading them out over too (which saves us an apt update in the build process), and I think is clearer from a documentation pov too.

**Review by bloodearnest — APPROVED** [review]

**Pull request merged**